File: Blind 75/04 - Stack/20-valid-parentheses.py
'''
Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

1. Open brackets must be closed by the same type of brackets.
2. Open brackets must be closed in the correct order.
3. Every close bracket has a corresponding open bracket of the same type.
'''

from time import time


class Solution:
    def isValid(self, s: str) -> bool:
        pair = {'(': ')', '{': '}', '[': ']'}
        stack = []
        for c in s:
            if c in pair:
                stack.append(c)
                continue
            if len(stack) == 0 or c != pair[stack[-1]]:
                # Invalid if adding closing bracket to empty stack or if mismatched
                return False
            stack.pop()

        return len(stack) == 0

    def reference(self, s: str) -> bool:
        bracket_map = {')': '(', ']': '[', '}': '{'}
        stack = []

        for c in s:
            if c not in bracket_map:
                stack.append(c)
                continue
            if not stack or stack[-1] != bracket_map[c]:
                return False
            stack.pop()

        return not stack

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.isValid(case))
                else:
                    self.isValid(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        '()',
        '()[]{}',
        '(]',
        # Additional
        '(()',
        '[()',
        '(])',
    ]
    test.quantify(test_cases)

File: Blind 75/04 - Stack/JavaScript/20-valid-parentheses.js
const { performance } = require("perf_hooks");

class Solution {
  method(s) {
    const table = {
      ")": "(",
      "}": "{",
      "]": "[",
    };

    let stack = [];

    for (const c of s) {
      if (!(c in table)) {
        stack.push(c);
        continue;
      }
      if (stack.length === 0 || stack.slice(-1)[0] !== table[c]) return false;
      stack.pop();
    }

    return true;
  }

  reference(s) {
    if (!s) return false;
    let closeMap = {
      "}": "{",
      "]": "[",
      ")": "(",
    };
    let stack = [];
    for (const str of s) {
      if (str in closeMap) {
        if (stack.length !== 0 && stack.slice(-1)[0] === closeMap[str]) {
          stack.pop();
        } else {
          return false;
        }
      } else {
        stack.push(str);
      }
    }
    return stack.length === 0;
  }

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = ["()", "()[]{}", "(]"];
test.quantify(testCases);
